File: PokemonJava/src/test/java/PokemonTest.java
import Exceptions.UnknownMoveException;
import Pokemon.Pokemon;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class PokemonTest {

    private Pokemon engine;

    @BeforeEach
    public void initEngine() {
        this.engine = new Pokemon();
    }

    @Test
    public void TestGivenCases() throws UnknownMoveException {
        String moves = "E";

        int expectedCount = 2;

        int result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);

        initEngine();

        moves = "NESO";

        expectedCount = 4;

        result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);

        initEngine();

        moves = "NSNSNSNSNS";

        expectedCount = 2;

        result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);
    }

    @Test
    public void TestCustom() throws UnknownMoveException {
        String moves = "NSSN";

        int expectedCount = 3;

        int result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);

        initEngine();

        moves = "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN";

        expectedCount = 200;

        result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);

    }

    @Test
    public void TestEmpty() throws UnknownMoveException {
        String moves = "";

        int expectedCount = 1;

        int result = engine.gottaCatchEmAll(moves);

        Assertions.assertEquals(expectedCount, result);


    }

    @Test()
    public void TestForUnknownMove() {

        String moves = "Isto deve dar erro (hopefully)";


        Assertions.assertThrows(UnknownMoveException.class, () -> engine.gottaCatchEmAll(moves));

    }
}

File: PokemonGo/pokemon_test.go
package main

import (
	"testing"
)

func TestGivenTestCases(t *testing.T) {

	moves := "E"
	expectedCount := 2

	count, err := GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)

	moves = "NESO"
	expectedCount = 4

	count, err = GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)

	moves = "NSNSNSNSNS"
	expectedCount = 2

	count, err = GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)

}

func TestCustom(t *testing.T) {

	moves := "NSSN"
	expectedCount := 3

	count, err := GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)

	moves = "NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN"
	expectedCount = 200

	count, err = GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)
}

func TestForError(t *testing.T) {

	moves := "Isto deve dar erro (hopefully)"
	expectedCount := 0

	count, err := GottaCatchEmAll(moves)

	checkForErrorHandling(err, count, expectedCount, moves, t)
}

//Assumi que se nenhuma jogada for iserida, o Ash apanha so o Pokemon da casa inicial
func TestEmpty(t *testing.T) {

	moves := ""
	expectedCount := 1

	count, err := GottaCatchEmAll(moves)

	checkForNormalExecution(err, count, expectedCount, moves, t)
}

func checkForNormalExecution(err error, count int, expectedCount int, moves string, t *testing.T) {
	if err != nil {
		t.Error(err)
	} else if count != expectedCount {
		t.Errorf("Error for move '%s': expected %d got %d", moves, expectedCount, count)
	}
}

func checkForErrorHandling(err error, count int, expectedCount int, moves string, t *testing.T) {

	if err == nil {
		t.Error("Error was not caught")
	} else if count != expectedCount {
		t.Errorf("Error for move '%s': expected %d got %d", moves, expectedCount, count)
	}
}
